src/buffer/h264.rs

```diff
@@ -304,3 +304,425 @@ impl IQMatrixBufferH264 {
+                mb_adaptive_frame_field_flag,
+                seq_scaling_matrix_present_flag,
+                direct_8x8_inference_flag,
+                log2_max_frame_num_minus4,
+                pic_order_cnt_type,
+                log2_max_pic_order_cnt_lsb_minus4,
+                delta_pic_order_always_zero_flag,
+            );
+
+        Self(bindings::_VAEncSequenceParameterBufferH264__bindgen_ty_1 {
+            bits: bindings::_VAEncSequenceParameterBufferH264__bindgen_ty_1__bindgen_ty_1 {
+                _bitfield_align_1: Default::default(),
+                _bitfield_1,
+                __bindgen_padding_0: Default::default(),
+            },
+        })
+    }
+
+    pub fn inner(&self) -> &bindings::_VAEncSequenceParameterBufferH264__bindgen_ty_1 {
+        &self.0
+    }
+}
+
+#[derive(Default)]
+pub struct H264VuiFields(bindings::_VAEncSequenceParameterBufferH264__bindgen_ty_2);
+
+impl H264VuiFields {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        aspect_ratio_info_present_flag: u32,
+        timing_info_present_flag: u32,
+        bitstream_restriction_flag: u32,
+        log2_max_mv_length_horizontal: u32,
+        log2_max_mv_length_vertical: u32,
+        fixed_frame_rate_flag: u32,
+        low_delay_hrd_flag: u32,
+        motion_vectors_over_pic_boundaries_flag: u32,
+    ) -> Self {
+        let _bitfield_1 =
+            bindings::_VAEncSequenceParameterBufferH264__bindgen_ty_2__bindgen_ty_1::new_bitfield_1(
+                aspect_ratio_info_present_flag,
+                timing_info_present_flag,
+                bitstream_restriction_flag,
+                log2_max_mv_length_horizontal,
+                log2_max_mv_length_vertical,
+                fixed_frame_rate_flag,
+                low_delay_hrd_flag,
+                motion_vectors_over_pic_boundaries_flag,
+                Default::default(),
+            );
+
+        Self(bindings::_VAEncSequenceParameterBufferH264__bindgen_ty_2 {
+            bits: bindings::_VAEncSequenceParameterBufferH264__bindgen_ty_2__bindgen_ty_1 {
+                _bitfield_align_1: Default::default(),
+                _bitfield_1,
+            },
+        })
+    }
+}
+
+#[derive(Default)]
+pub struct H264EncFrameCropOffsets {
+    pub left: u32,
+    pub right: u32,
+    pub top: u32,
+    pub bottom: u32,
+}
+
+impl H264EncFrameCropOffsets {
+    pub fn new(
+        frame_crop_left_offset: u32,
+        frame_crop_right_offset: u32,
+        frame_crop_top_offset: u32,
+        frame_crop_bottom_offset: u32,
+    ) -> Self {
+        Self {
+            left: frame_crop_left_offset,
+            right: frame_crop_right_offset,
+            top: frame_crop_top_offset,
+            bottom: frame_crop_bottom_offset,
+        }
+    }
+}
+
+pub struct EncSequenceParameterBufferH264(Box<bindings::VAEncSequenceParameterBufferH264>);
+
+impl EncSequenceParameterBufferH264 {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
```

Comment:
For these complex types that have many fields, many of which we will want to leave to their default value, I wonder if using the builder pattern would not be simpler?

I.e. the `new` method only takes parameters that we know must be specified all the time, and other parameters are provided by homonymous methods that take `Self` and return a modified `Self`.

Please ignore if it turns out that we always need to specify meaningful values for all the parameters, but the example suggests that we don't. I'm also ok with delegating that to a later CL if the answer is unsure.

Reply:
In general I like the idea of using builder pattern for SPS and PPS structures. On top of that currently using `::new` it's hard to tell what parameter the value is for without editor with parameter hints. However in cros-codecs we will most likely map other structure to those (eg. `cros_codecs::codec::h264::parser::Sps` to `EncSequenceParameterBufferH264`) and set all fields anyway. So I wonder if the better place for this would be in cros-codecs.